src/main/cpp/src/get_json_object.hpp

```diff
@@ -0,0 +1,1191 @@
+  //       if (json_token::ERROR == p.get_current_token()) { return false; }
+
+  //       return path_evaluator::evaluate_path(p, g, style, path_ptr + 1, path_size - 1);
+  //     } else {
+  //       return false;
+  //     }
+  //   }
+  //   // case (FIELD_NAME, Wildcard :: xs)
+  //   // case path 11
+  //   else if (json_token::FIELD_NAME == token &&
+  //            path_match_element(path_ptr, path_size, path_instruction_type::WILDCARD)) {
+  //     p.next_token();
+  //     // JSON validation check
+  //     if (json_token::ERROR == p.get_current_token()) { return false; }
+
+  //     return path_evaluator::evaluate_path(p, g, style, path_ptr + 1, path_size - 1);
+  //   }
+  //   // case _ =>
+  //   // case path 12
+  //   else {
+  //     if (!p.try_skip_children()) { return false; }
+  //     return false;
+  //   }
+  // }
+
+  /**
+   *
+   * This function is rewritten from above commented recursive function.
+   * this function is equivalent to the above commented recursive function.
+   */
+  static __device__ bool evaluate_path(json_parser<>& p,
+                                       json_generator<>& root_g,
+                                       write_style root_style,
+                                       path_instruction const* root_path_ptr,
+                                       int root_path_size)
+  {
+    // manually maintained context stack in lieu of calling evaluate_path recursively.
+    struct context {
+      // current token
+      json_token token;
+
+      // which case path that this task is from
+      int case_path;
+
+      // used to save current generator
+      json_generator<> g;
+
+      write_style style;
+      path_instruction const* path_ptr;
+      int path_size;
+
+      // is this context task is done
+      bool task_is_done = false;
+
+      // whether written output
+      // if dirty > 0, indicates success
+      int dirty = 0;
+
+      // for some case paths
+      bool is_first_enter = true;
+
+      // used to save child JSON generator for case path 8
+      json_generator<> child_g;
+
+      __device__ context()
+        : token(json_token::INIT),
+          case_path(-1),
+          g(json_generator<>()),
+          style(write_style::raw_style),
+          path_ptr(nullptr),
+          path_size(0)
+      {
+      }
+
+      __device__ context(json_token _token,
+                         int _case_path,
+                         json_generator<> _g,
+                         write_style _style,
+                         path_instruction const* _path_ptr,
+                         int _path_size)
+        : token(_token),
+          case_path(_case_path),
+          g(_g),
+          style(_style),
+          path_ptr(_path_ptr),
+          path_size(_path_size)
+      {
+      }
+
+      __device__ context& operator=(context const&) = default;
+    };
+
+    // path max depth limitation
+    constexpr int max_path_depth = 32;
+
+    // stack
+    context stack[max_path_depth];
```

Comment:
I'm not sure about  this declaration but feel really nervous about it. Typically, a GPU thread has very limited resources (memory/registers etc). Reserving a large array in a thread will stress on the resources, which results in fewer number of threads per block and reduces performance. Probably the better way is to declare this as a shared memory array instead. I'm not 100% sure until we can run profiling on this.

Reply:
In addition, the code below also has a lot of if/else branches without any thread sync. Thus, thread divergence may occur which causes further slowdown. I have little idea about what can we improve for that huge `while` loop :cry:

Reply:
I filed a following on perf issue:
https://github.com/NVIDIA/spark-rapids-jni/issues/1907